engine/exception.go

```diff
@@ -290,15 +305,24 @@ func (r resource) Term() Term {
 	return resourceAtoms[r]
 }
 
+// ResourceError creates a new resource error exception.
+func ResourceError(resource Term, env *Env) Exception {
+	return Exception{term: atomError.Apply(atomResourceError.Apply(resource), env.Resolve(varContext))}
+}
+
 // resourceError creates a new resource error exception.
 func resourceError(resource resource, env *Env) Exception {
-	// We can't call renamedCopy() since it can lead th resource_error(memory).
-	return Exception{term: atomError.Apply(atomResourceError.Apply(resource.Term()), env.Resolve(varContext))}
+	return ResourceError(resource.Term(), env)
+}
+
+// SyntaxError creates a new syntax error exception.
+func SyntaxError(err error, env *Env) Exception {
```

Comment:
Shouldn't this be `func SyntaxError(error Term, env *Env) Exception`?

We don't need an actual Go error to throw a syntax_error. For example, let `number_of_as/2` be a predicate which takes a sequence of `a` as an atom and returns the number of runes. If the atom contains a rune that is not `a`, it's a syntax error. This syntax error isn't backed by a Go error.

```prolog
?- number_of_as('', N).
N = 0.
?- number_of_as('a', N).
N = 1.
?- number_of_as('aa', N).
N = 2.
?- number_of_as('aab', N).
2024/01/21 23:24:25 error(syntax_error(rune(b,2)),number_of_as/2)
```

```go
func NumberOfAs(vm *VM, input, number Term, k Cont, env *Env) *Promise {
	var in string
	switch i := env.Resolve(input).(type) {
	case Variable:
		return Error(InstantiationError(env))
	case Atom:
		in = i.String()
	default:
		return Error(typeError(validTypeAtom, i, env))
	}

	switch n := env.Resolve(number).(type) {
	case Variable:
		break
	case Integer:
		if n < 0 {
			return Error(domainError(validDomainNotLessThanZero, n, env))
		}
	default:
		return Error(typeError(validTypeInteger, n, env))
	}

	var c Integer
	for i, r := range in {
		switch r {
		case 'a':
			c++
		default:
			// Syntax error but no actual Go error here!
			return Error(SyntaxError(NewAtom("rune").Apply(NewAtom(string(r)), Integer(i)), env))
		}
	}

	return Unify(vm, number, c, k, env)
}
```

Reply:
Yes, absolutely. I agree with that.